chain/chain/src/chain.rs

```diff
@@ -1903,81 +1903,111 @@ impl Chain {
         Ok(chain_store_update.commit()?)
     }
 
+    pub fn catchup_blocks_step(
+        &mut self,
+        me: &Option<AccountId>,
+        sync_hash: &CryptoHash,
+        blocks_catch_up_state: &mut BlocksCatchUpState,
+        block_catch_up_scheduler: &dyn Fn(BlockCatchUpRequest),
+    ) -> Result<(), Error> {
+        for (queued_block, (saved_store_update, results)) in
+            blocks_catch_up_state.processed_blocks.drain()
+        {
+            match self.block_catch_up_postprocess(&queued_block, results, saved_store_update) {
+                Ok(_) => {
+                    let mut saw_one = false;
+                    for next_block_hash in self.store.get_blocks_to_catchup(&queued_block)?.clone()
+                    {
+                        saw_one = true;
+                        blocks_catch_up_state.pending_blocks.push(next_block_hash.clone());
+                    }
+                    if saw_one {
+                        assert_eq!(
+                            self.runtime_adapter.get_epoch_id_from_prev_block(&queued_block)?,
+                            blocks_catch_up_state.epoch_id
+                        );
+                    }
+                    blocks_catch_up_state.done_blocks.push(queued_block);
+                }
+                Err(_) => {
+                    error!("Error processing block during catch up, retrying");
+                    blocks_catch_up_state.pending_blocks.push(queued_block);
+                }
+            }
+        }
+
+        for pending_block in blocks_catch_up_state.pending_blocks.drain(..) {
+            let block = self.store.get_block(&pending_block)?.clone();
+            let prev_block = self.store.get_block(block.header().prev_hash())?.clone();
+
+            let mut chain_update = self.chain_update();
+            let work = chain_update.apply_chunks_preprocessing(
+                me,
+                &block,
+                &prev_block,
+                ApplyChunksMode::CatchingUp,
+            )?;
+            blocks_catch_up_state
+                .scheduled_blocks
+                .insert(pending_block, chain_update.save_store_update());
+            block_catch_up_scheduler(BlockCatchUpRequest {
```

Comment:
I'm not familiar with the Arbiter interface. What kind of guarantee does it make for the ordering of sending and receiving messages? It's very important here that the ordering by which `pending_blocks` get processed are correct since blocks must be processed sequentially.

Reply:
Actually I think the for loop here is confusing and may be incorrect. I don't think more than one blocks should be send to `SyncJobActor` at the same time, the request will simply fail because the information needed for the later block is not ready in the database yet. Processed block information will only be stored to the database in `block_catch_up_postprocess`, and a block must be post-processed before the next block can be processed. 


A more correct and simpler logic would work like this. There should be a queue for pending blocks, and an `Option` for processed block since there will be at most one processed block at a time. At every step, this function checks if there is an processed block, if so, it postprocesses that block and sends the next block from the pending queue to be processed.  At any given point, there should be at most one block being processed

Reply:
We only send more than one block per sync_hash if there is a fork. We do not send any block for pre-processing unless we post-processed its predecessor. So we do not care about any guarantees on order of message processing

Reply:
Ah I forgot about forks, yeah now the implementation makes sense to me, sorry for the confusion